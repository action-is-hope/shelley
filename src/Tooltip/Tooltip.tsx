import React, { forwardRef, useState } from "react";
import { useTooltipTriggerState } from "react-stately";
import { mergeProps, useTooltip, useTooltipTrigger } from "react-aria";

interface TooltipProps extends React.HTMLAttributes<HTMLDivElement> {
  children: any;
  state: any;
}

function Tooltip(props: TooltipProps, ref: React.Ref<HTMLDivElement>) {
  let { tooltipProps } = useTooltip(props, props.state);

  return (
    <span
      style={{
        position: "absolute",
        left: "5px",
        top: "100%",
        maxWidth: 150,
        marginTop: "10px",
        backgroundColor: "white",
        color: "black",
        padding: "5px",
        border: "1px solid gray",
      }}
      ref={ref}
      {...mergeProps(props, tooltipProps)}
    >
      {props.children}
    </span>
  );
}

const _Tooltip = forwardRef(Tooltip);
export { _Tooltip as Tooltip };

// function TooltipButton(props) {
//   let state = useTooltipTriggerState(props);
//   let ref = React.useRef(null);

//   // Get props for the trigger and its tooltip
//   let { triggerProps, tooltipProps } = useTooltipTrigger(props, state, ref);

//   return (
//     <span style={{ position: 'relative' }}>
//       <button
//         ref={ref}
//         {...triggerProps}
//         style={{ fontSize: 18 }}
//         onClick={() => alert('Pressed button')}
//       >
//         {props.children}
//       </button>
//       {state.isOpen && (
//         <Tooltip state={state} {...tooltipProps}>{props.tooltip}</Tooltip>
//       )}
//     </span>
//   );
// }

// <TooltipButton tooltip="Edit">‚úèÔ∏è</TooltipButton>
// <TooltipButton tooltip="Delete">üöÆ</TooltipButton>
