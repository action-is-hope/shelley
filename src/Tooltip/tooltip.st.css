@namespace "Tooltip";

.root {
  position: absolute;
  left: 5px;
  top: 100%;
  max-width: 150;
  margin-top: 10px;
  background-color: white;
  color: black;
  padding: 5px;
  border: 1px solid gray;
}
