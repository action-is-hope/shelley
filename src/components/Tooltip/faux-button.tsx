import React from "react";

const h = React.createElement;

/**
 * Accessible "button" usurper.
 * Establishes the correct role, tabindex, and key/mouse interaction if interactive.
 * Buttons activate on space and enter.
 *
 * NOTE: Gives 'disabled' attribute if interaction not provided.
 *
 * @see https://www.w3.org/TR/wai-aria-practices-1.1/#button
 * @see https://www.w3.org/TR/wai-aria-practices-1.1/examples/button/button.html
 */
interface FauxButtonProps extends React.HtmlHTMLAttributes<HTMLButtonElement> {
  /** The DOM element for the component */
  tag: string;
  className?: string;
  /** The action to take on activation. Happens on click and space or enter keys. */
  onInteraction?: () => any;
  /** HTML disabled attribute */
  disabled?: boolean;
}

/** Handle the keyboard activation same as a button */
function handleKeyActivation(cb: () => any) {
  return (event: any) => {
    switch (event.key) {
      case "Enter":
      case " ":
        event.preventDefault();
        cb();
        return;
    }
  };
}

export const FauxButton: React.SFC<FauxButtonProps> = ({
  tag,
  className,
  onInteraction,
  disabled,
  children,
  ...rest
}) =>
  h(
    tag,
    {
      ...rest,
      role: "button",
      tabIndex: "0",
      className,
      // Disabled if set, or if interaction non-existent
      disabled: disabled ? true : onInteraction ? null : true,
      onClick: onInteraction,
      onKeyDown: onInteraction && handleKeyActivation(onInteraction)
    },
    children
  );
