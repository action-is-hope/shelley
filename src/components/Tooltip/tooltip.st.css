/* Contain the positioning of the tooltip */
.root {
  display: inline-block;
  position: relative;
}

/* Hide tooltip by default */
.tooltip {
  display: none;
  position: absolute;
  top: 100%;
  padding: 1em;
  transform: translateY(0.25em);
  background: #000;
  color: #fff;
}

.button {
  text-decoration: underline;
  text-decoration-style: dotted;
}

/* Display if button is hovered/focused */
.button:hover + .tooltip,
.button:focus + .tooltip {
  display: block;
}
