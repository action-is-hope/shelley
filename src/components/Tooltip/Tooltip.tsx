import React from "react";
import { FauxButton } from "./faux-button";
// import "./styles.css";
import style from "./tooltip.st.css";

export type TooltipPurpose = "primary" | "descriptive";

export interface TooltipProps {
  /** Whether the Tooltip labels or elaborates on the element */
  purpose: TooltipPurpose;
  /** Render callback for content */
  renderContent?: React.ReactNode;
  /** Render callback for tooltip */
  renderTooltip: React.ReactNode;
}

/* Component that provides a Tooltip as the primary label.
 * Useful in cases where the content should be always accessible to screen readers
 * and on interaction for mouse and keyboard.
 * An example would be making a better `abbr` element.
 * Based on https://inclusive-components.design/tooltips-toggletips/, particularly
 * the "Tooltip as Primary Label" use-case.
 *
 * NOTE: the aria-labelledby of the element supersedes the text content of renderContent,
 * in the same way that `title` supersedes text content in abbr.
 */
export const Tooltip = ({
  purpose,
  renderContent,
  renderTooltip,
  children,
  ...rest
}: TooltipProps) => (
  <RandomId>
    {id => (
      <span {...style(style.root, {}, rest)}>
        {/* Using FauxButton to avoid having to pass on resets from consumers */}
        <FauxButton
          tag="span"
          className={style.button}
          aria-labelledby={purpose === "primary" ? id : undefined}
          aria-describedby={purpose === "descriptive" ? id : undefined}
        >
          {children}
        </FauxButton>
        <div className={style.tooltip} role="tooltip" id={id}>
          {renderTooltip}
        </div>
      </span>
    )}
  </RandomId>
);

// Utils
/** Render-prop component that passes a random id */
interface RandomIdProps {
  /** Custom prefix for the id */
  prefix?: string;
  children: (id: string) => React.ReactNode;
}

class RandomId extends React.Component<RandomIdProps> {
  labelId: string;

  constructor(props: RandomIdProps) {
    super(props);

    // Allocate a random id on init
    this.labelId =
      this.props.prefix +
      Math.random()
        .toString(36)
        .substr(2, 9);
  }

  render() {
    const { children } = this.props;
    return children(this.labelId);
  }
}
