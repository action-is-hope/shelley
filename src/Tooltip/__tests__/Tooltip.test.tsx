import { Tooltip } from "../Tooltip";
import renderer from "react-test-renderer";

const tooltipText = "Tooltip text";
const tooltipStateMock = {
  isOpen: false,
  close: jest.fn(),
  open: jest.fn(),
};

describe("Tooltip", () => {
  it("renders with given text", () => {
    const tree = renderer
      .create(
        <Tooltip state={tooltipStateMock} id="test-tooltip">
          {tooltipText}
        </Tooltip>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders with custom state", () => {
    const customState = {
      ...tooltipStateMock,
      isOpen: true,
    };
    const tree = renderer
      .create(
        <Tooltip state={customState} id="test-tooltip-open">
          {tooltipText}
        </Tooltip>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
