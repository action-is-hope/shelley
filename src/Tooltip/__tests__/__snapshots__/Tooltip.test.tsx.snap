// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip renders with custom state 1`] = `
<span
  className="Tooltip__Test__root"
  id="test-tooltip-open"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onTouchStart={[Function]}
  role="tooltip"
  state={
    {
      "close": [MockFunction],
      "isOpen": true,
      "open": [MockFunction],
    }
  }
>
  Tooltip text
</span>
`;

exports[`Tooltip renders with given text 1`] = `
<span
  className="Tooltip__Test__root"
  id="test-tooltip"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onTouchStart={[Function]}
  role="tooltip"
  state={
    {
      "close": [MockFunction],
      "isOpen": false,
      "open": [MockFunction],
    }
  }
>
  Tooltip text
</span>
`;
